crates/wasmtime/src/runtime/instance.rs

```diff
@@ -259,7 +259,7 @@ impl Instance {
         store.bump_resource_counts(module)?;
 
         // Allocate the GC heap, if necessary.
-        if store.engine().features().gc_types() {
+        if store.engine().features().gc_types() && module.env_module().needs_gc_heap {
```

Comment:
I think here we might be able to entirely remove the first part of the condition?